### PR TITLE
[cmd/opampsupervisor] Persist collector remote config & telemetry set…

### PR DESCRIPTION
**Pull request by srikanthccv**
…tings

**Description:** <Describe what has changed.>
Persists the configs contained in the AgentRemoteConfig and ConnectionSettings messages to be used on startup before any messages have been received from the OpAMP server. If the capability corresponding to the message for the config is disabled, then the config is not merged into the Collector's effective configuration

**Link to tracking Issue:** Part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21078

**Testing:**

Added e2e test

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 1]
```diff
@@ -784,9 +837,38 @@ func (s *Supervisor) Shutdown() {
...
+	if s.config.Storage == nil {
+		return nil
+	}
+
+	cfg, err := proto.Marshal(config)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(filepath.Join(s.config.Storage.Directory, lastRecvRemoteConfigFile), cfg, 0644)
```
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor.go:850]: Should we make the permissions on this file more restrictive, e.g. `0600`?

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 2]
```diff
@@ -460,6 +465,48 @@ func (s *Supervisor) loadAgentEffectiveConfig() {
 	}
 
 	s.effectiveConfig.Store(string(effectiveConfigBytes))
+
+	if s.config.Capabilities.AcceptsRemoteConfig != nil &&
+		*s.config.Capabilities.AcceptsRemoteConfig &&
+		s.config.Storage != nil {
+		// Try to load the last received remote config if it exists.
+		lastRecvRemoteConfig, err := os.ReadFile(filepath.Join(s.config.Storage.Directory, lastRecvRemoteConfigFile))
```
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor.go:473]: I realize it would make the implementation a little more complex, but what are your thoughts on individually persisting each config file passed in the `RemoteConfig` message instead of the message itself? That way the files would be available to be easily read by an operator if necessary. Granted, I'm not sure how useful that would be in practice.

I think persisting them separately could also make it easier in the future to separately pass the files to the Collector like `otelcol --config config1.yaml --config config2.yaml [...]`. This would let the Collector do the merging instead of the Supervisor doing it.

**Review by evan-bradley — COMMENTED** [review]
I know it's marked as a draft, but overall looks good to me right now.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 2 above)
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:473]: That's a good point, I agree.

Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:473]: Can we do this after addressing https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30622? I started working on the change and soon ended up changing the effective config report part as well which makes this change more complex.

Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:473]: That's good with me.

**Comment by BinaryFissionGames** [comment]
@srikanthccv Are you still working on this? I'd be happy to help push this forward.

**Comment by srikanthccv** [comment]
Hi @BinaryFissionGames, I am on vacation till 1st Apr. I will revive this after I come back.

**Review by codeboten — COMMENTED** [review]
Thanks for the continued work @srikanthccv, please resolve the conflicts. @evan-bradley if you could review, would be great to move this forward.

File: cmd/opampsupervisor/e2e_test.go [hunk 3]
```diff
@@ -518,3 +518,48 @@ func TestSupervisorOpAMPConnectionSettings(t *testing.T) {
...
+					"": {Body: cfg.Bytes()},
+				},
+			},
+			ConfigHash: hash,
+		},
+	})
+
+	require.Eventually(t, func() bool {
+		// ensure that the config file was written to the storage directory
+		_, err := os.Stat(path.Join(tempDir, "last_recv_remote_config.yaml"))
```
Comment (evan-bradley) [inline cmd/opampsupervisor/e2e_test.go:562]: Since this is a high-level integration/E2E test, could we instead test that if we stop the OpAMP server, then restart the Supervisor, that we get a Collector that starts with the last-received remote config?

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 4]
```diff
@@ -49,6 +51,9 @@ var (
 
 	//go:embed templates/owntelemetry.yaml
 	ownTelemetryTpl string
+
+	lastRecvRemoteConfigFile     = "last_recv_remote_config.yaml"
+	lastRecvOwnMetricsConfigFile = "last_recv_own_metrics_config.yaml"
```
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor.go:56]:
```suggestion
	lastRecvRemoteConfigFile     = "last_recv_remote_config.proto"
	lastRecvOwnMetricsConfigFile = "last_recv_own_metrics_config.proto"
```
These are proto files, right?

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 5]
```diff
@@ -561,6 +567,48 @@ func (s *Supervisor) loadAgentEffectiveConfig() {
...
+			config := &protobufs.AgentRemoteConfig{}
+			err = proto.Unmarshal(lastRecvRemoteConfig, config)
+			if err != nil {
+				s.logger.Error("Cannot parse last received remote config", zap.Error(err))
+			} else {
+				s.remoteConfig = config
+			}
+		}
+	} else {
+		s.logger.Debug("Remote config is not supported")
```
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor.go:586]:
```suggestion
		s.logger.Debug("Remote config is not supported, will not attempt to load config from file")
```
We should probably be more specific about the context of the log.

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 6]
```diff
@@ -561,6 +567,48 @@ func (s *Supervisor) loadAgentEffectiveConfig() {
...
+			set := &protobufs.TelemetryConnectionSettings{}
+			err = proto.Unmarshal(lastRecvOwnMetricsConfig, set)
+			if err != nil {
+				s.logger.Error("Cannot parse last received own metrics config", zap.Error(err))
+			} else {
+				s.setupOwnMetrics(context.Background(), set)
+			}
+		}
+	} else {
+		s.logger.Debug("Own metrics is not supported")
```
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor.go:604]:
```suggestion
		s.logger.Debug("Own metrics is not supported, will not attempt to load config from file")
```
Same as above.

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 7]
```diff
@@ -626,39 +674,45 @@ func (s *Supervisor) setupOwnMetrics(_ context.Context, settings *protobufs.Tele
 // 2) the own metrics config section
 // 3) the local override config that is hard-coded in the Supervisor.
 func (s *Supervisor) composeEffectiveConfig(config *protobufs.AgentRemoteConfig) (configChanged bool, err error) {
```
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor.go:676]: I generally agree with these changes, but could we do them in a separate PR? I'd like to make sure we're all on the same page around the semantics of the ConfigMap keys in the RemoteConfig message, and plus that way we can add tests etc. without making this PR too large.

File: cmd/opampsupervisor/supervisor/supervisor_test.go [hunk 8]
```diff
@@ -55,6 +59,6 @@ func Test_composeEffectiveConfig(t *testing.T) {
 	require.NoError(t, err)
 	expectedConfig = bytes.ReplaceAll(expectedConfig, []byte("\r\n"), []byte("\n"))
 
-	require.True(t, configChanged)
+	require.False(t, configChanged)
```
Comment (evan-bradley) [inline cmd/opampsupervisor/supervisor/supervisor_test.go:62]: I know the `composeEffectiveConfig` function was updated, but it's not immediately obvious to me why this changed. Do you know why this is different now?

**Review by evan-bradley — COMMENTED** [review]
Sorry for the delay, thanks for keeping this going.

File: cmd/opampsupervisor/e2e_test.go (hunk 3 above)
Comment (srikanthccv) [reply cmd/opampsupervisor/e2e_test.go:562]: This is a good idea. I will update the test.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 4 above)
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:56]: We are marshalling the protos and writing that content to file.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 7 above)
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:676]: There were no functional changes here. I mainly added the nil checks here. If I remove the nil checks, then rest of the code remains the same.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 4 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:56]: Thinking again, maybe `.proto` isn't the right file extension here. But since we're marshalling them, shouldn't they be `.dat` or `.protobuf` instead of `.yaml` since they're not YAML files?

File: cmd/opampsupervisor/supervisor/supervisor_test.go (hunk 8 above)
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor_test.go:62]: The offered config update in line 51 is empty and doesn't change anything. I forgot why the current logic is returning true but let me check and get back. I think the updated behaviour is correct.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 4 above)
Comment (srikanthccv) [reply cmd/opampsupervisor/supervisor/supervisor.go:56]: Right, you are correct.

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 7 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:676]: My bad, not sure what I saw here. This is good with me, thanks.

File: cmd/opampsupervisor/supervisor/supervisor_test.go (hunk 8 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor_test.go:62]: Thanks for digging into it. I think we should probably update the test since that flow involves more logic.

**Comment by evan-bradley** [comment]
@srikanthccv FYI it looks like some conflicts were introduced by other Supervisor PRs.

**Review by evan-bradley — APPROVED** [review]
